### ISSUE TITLE
Implement an incremental backoff strategy for handling errors when syncing a parent

### ISSUE THREAD

**Issue by psFried** [psFried/roperator#3]
The current behavior uses fixed delays when an error is encountered while syncing a parent. We should implement a backoff strategy so that re-tries can happen more quickly at first, and then move to longer delays.

**Comment by cassandracomar** [comment]
I'll try and take a stab at a PR for this. New to rust so it may take me a little while. 

Initial thoughts are to add an enum like
```
enum ResyncPeriod {
    Resync(Optional<Box<Error>>),                                         // just use standard exponential back off
    ResyncAfter(std::time::Duration, Optional<Box<Error>>),  // exponential back off with an initial delay 
    ResyncNever(Optional<Box<Error>>),                                // no automatic resyncs until an update occurs
                                                                                                // include an optional error so caller can track
}
```
and allow it as an optional return parameter on the sync function. After that, I think it's as simple as updating the code that decides when to resync to respect these cases.

Does that sound about right?

**Comment by psFried** [comment]
Hey thanks a lot for offering to take this. I'd appreciate the help.
I actually started to implement something similar on the [error-handling branch](https://github.com/psFried/roperator/compare/error-handling?expand=1) and had a few realizations during the process.

- The `Handler` is really only handling its own errors, not errors that occur within the rest of the framework (e.g. http errors from the k8s api server). Given that, we don't necessarily _need_ to add anything to the SyncResponse apart from a `Option<Duration>` that represents the desire to re-sync at some point in the future. If that field is `Some`, then we schedule a resync for the given period in the future, and if it's None then we continue as usual. The `Handler::sync` function already returns a `Result` that can be an `Error`, so for cases where the "default" backoff strategy is desired, users can just return the error that way. 
- Typically, users will want to include information about the error in the `status` of the parent, so it makes sense for error handling to happen before the status is returned in the response.
- Users may _also_ want to schedule a resync even when no error occurred. This would be useful in the case that the operator is syncing some state in an external system.
- If we add just the `Option<Duration>` field to the response, then we could allow for multiple different error handling strategies by introducing new traits/impls that all just implement `Handler`. For example, we could have an `ErrorHandlingHandler` that has pre-defined behaviors for backoff and for including error information on the status.

Some of this is already implemented on the error-handling branch. The resync field is added to the response, and the controller is properly scheduling resyncs for the future if desired. What remains to be done is to implement the backoff logic in the controller, and to add some sort of Handler impl that has reasonable pre-defined error-handling behaviors.

I kind of went back and forth on whether to use an `Option<Duration>` vs some sort of `Resync` struct/enum that has the same semantics but with perhaps a more obvious api. So we may want to change that, too.

Anyway, I could merge the error-handling branch to master today, to make it easier for us to collaborate on this. Sorry for the rambling response, but let me know if there's anything that doesn't make sense yet, or if you think of an alternative that might make more sense.

**Comment by cassandracomar** [comment]
that all makes sense. i was suggesting the broader api because I think there's an additional case you can't cover well with an `Option`. sometimes, you hit a case where you know there's no point to resyncing state because something fundamentally doesn't line up. you could keep logging errors about the condition but the state either needs to be manually fixed or something outside the scope of the operator needs to change. for cases like these, it helps to have a method to inform the Handler to just not resync unless an update (or other) event actually hits the queue. Something that says "I know I can't make progress" but that is distinguished from a successful sync.

I see your point about just using the status to record errors and I think that makes sense.

so I guess I'm suggesting something like
```
enum ResyncPeriod {
    Resync(),                          // continue on as normal
    ResyncAfter(std::time::Duration),  // resync after specified delay
    ResyncNever(),                     // don't bother until something changes
}
```
or perhaps, more simply, `Result<Option<Duration>, ()>`, where the Result wrapper denotes whether or not to resync with an optional `Duration`. `()` can also be replaced with an actual `Error` type for recording as an event on the resource.

(a well-supported way to tag structured events on the resource without having to make explicit api calls from the user code is something it seems like a lot of operator stuff, even in golang, just lacks).

I'll check out your branch and see about adding in the backoff.

**Comment by psFried** [comment]
@cassandracomar  I think I'm understanding where you're coming from. You're proposing a single code path for error handling. As things are right now, there's two separate ways that you could handle any error:

- Just return an error from the `sync` function, which will mean that the default error handling logic will apply. Right now, that's to delay for a hard-coded 10 seconds, and then retry, but in the future this should probably be an incremental backoff.
- Handle the error within the `sync` function, which allows you to set a status on the parent with info on the error. Then you can choose whether to set the `resync` field or not. If you do, then we'll retry after the given period, if not then we'll only retry when something gets updated.

That's the _current_ state of affairs, anyway. I think what you're getting at is that the idea would be to unify those two paths, if we can, and I think that's a great idea. 

One thought about how to do that would be to just change the function signature of `Handler::sync` so that it returns a plain `SyncResponse` instead of a `Result<SyncResponse, Error>`. If we did that, then we could just move _all_ of the error handling into the `Handler` impl. Maybe something kind of like:

```rust
pub trait Handler {
  fn sync(req: &SyncRequest) -> SyncResponse;
}

struct DefaultHandler {
  sync_fun: &fn(&SyncRequest) -> Result<SyncResponse, Error>,
  handle_error: &fn(&SyncRequest, Error) -> (Value, Option<Duration>),
}

impl Handler for SomeDefaultHandler {
  fn sync(req: &SyncRequest) -> SyncResponse {
    match self.sync_fun(req) {
      Ok(resp) -> resp,
      Err(err) -> {
        let (status, resync) = self.handle_error(req, err);
        SyncResponse { status, resync, children: Vec::new() }
      }
    }
  }
}
```

There's probably lots of room to improve the above, but that would be the general idea, anyway. The key thing being that `sync` can no longer return an error, so we'd now have one code path for error handling (that's 100% controlled by the user) instead of two separate paths. 

I'm kind of liking that idea. What do you think? Is that close to what you were thinking of?

**Comment by cassandracomar** [comment]
Yup! That does the trick!

**Comment by psFried** [comment]
@cassandracomar  Great! It sounds like some of the changes on #20 are still desirable, and some not. In the interest of easy collaboration, I'll just merge that PR to master, since master is collecting breaking changes for the 0.2 release anyway. Then I can work on the docs changes while you do the code?

**Comment by cassandracomar** [comment]
Sounds good! I have a first pass at the backoff implementation that I'm testing now and will start on the change to the `Handler` spec shortly.

**Comment by psFried** [comment]
#21 got this issue really close to being resolved, but there's still a few remaining tweaks. The work is being integrated on the `error-handling-trunk` branch until all the tests are passing, at which point we can merge to master. Please use `error-handling-trunk` as the base branch for any future PRs related to this.

**Comment by psFried** [comment]
The `handler-errors-redux` branch makes some additional changes beyond what what was discussed here. This is in PR #25 . That should finish off the remaining work to be done here.

**Comment by psFried** [comment]
Finally merged #25, which should close the loop on this.
